File: RideSharing.py
# ride_sharing_statistics.py

from collections import Counter

class StatisticsCalculator:
    def __init__(self, data):
        self.data = sorted(data)

    def calculate_mean(self):
        return sum(self.data) / len(self.data)

    def calculate_median(self):
        n = len(self.data)
        if n % 2 == 0:
            return (self.data[n//2 - 1] + self.data[n//2]) / 2
        else:
            return self.data[n//2]

    def calculate_mode(self):
        freq = Counter(self.data)
        max_count = max(freq.values())
        return [key for key, val in freq.items() if val == max_count]

if __name__ == "__main__":
    data = [2, 3, 2, 5, 3, 4, 2, 5, 5]
    calculator = StatisticsCalculator(data)

    print(f"Mean: {calculator.calculate_mean():.2f}")
    print(f"Median: {calculator.calculate_median():.2f}")
    print(f"Mode(s): {calculator.calculate_mode()}")

File: RideSharing.ml
(* ride_sharing_statistics.ml *)

let mean lst =
  let sum = List.fold_left ( +. ) 0.0 (List.map float_of_int lst) in
  sum /. float_of_int (List.length lst)

let median lst =
  let sorted = List.sort compare lst in
  let n = List.length sorted in
  if n mod 2 = 0 then
    let mid1 = List.nth sorted (n/2 - 1) in
    let mid2 = List.nth sorted (n/2) in
    (float_of_int (mid1 + mid2)) /. 2.0
  else
    float_of_int (List.nth sorted (n/2))

let mode lst =
  let count_map = List.fold_left (fun acc x ->
      if List.mem_assoc x acc then
        (x, (List.assoc x acc) + 1) :: (List.remove_assoc x acc)
      else
        (x, 1) :: acc
    ) [] lst in
  let max_count = List.fold_left (fun acc (_, count) -> max acc count) 0 count_map in
  List.filter (fun (_, count) -> count = max_count) count_map
  |> List.map fst

(* Test the functions *)
let () =
  let lst = [2; 3; 2; 5; 3; 4; 2; 5; 5] in
  Printf.printf "Mean: %.2f\n" (mean lst);
  Printf.printf "Median: %.2f\n" (median lst);
  Printf.printf "Mode(s): ";
  List.iter (fun x -> Printf.printf "%d " x) (mode lst);
  print_endline ""
